sdks/java/harness/src/main/java/org/apache/beam/fn/harness/state/MultimapSideInput.java

```diff
@@ -0,0 +1,89 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.fn.harness.state;
+
+import com.google.protobuf.ByteString;
+import java.io.IOException;
+import java.util.function.Supplier;
+import org.apache.beam.model.fnexecution.v1.BeamFnApi.StateRequest;
+import org.apache.beam.sdk.coders.Coder;
+import org.apache.beam.sdk.fn.stream.DataStreams;
+import org.apache.beam.sdk.transforms.Materializations.MultimapView;
+
+/**
+ * An implementation of a multimap side input that utilizes the Beam Fn State API to fetch values.
+ *
+ * <p>TODO: Support block level caching and prefetch.
+ */
+public class MultimapSideInput<K, V> implements MultimapView<K, V> {
+  private final BeamFnStateClient beamFnStateClient;
+  private final String sideInputId;
+  private final Coder<K> keyCoder;
+  private final Coder<V> valueCoder;
+  private final Supplier<StateRequest.Builder> partialRequestSupplier;
+
+  public MultimapSideInput(
+      BeamFnStateClient beamFnStateClient,
+      String sideInputId,
+      Coder<K> keyCoder,
+      Coder<V> valueCoder,
+      Supplier<StateRequest.Builder> partialRequestSupplier) {
+    this.beamFnStateClient = beamFnStateClient;
+    this.sideInputId = sideInputId;
+    this.keyCoder = keyCoder;
+    this.valueCoder = valueCoder;
+    this.partialRequestSupplier = partialRequestSupplier;
```

Comment:
<!--new_thread; commit:c3d73c64facc4d4670f203a8fda5fafa1fb8ef8b; resolved:0-->
Rather than passing around partialRequestSuppliers, it may be more straightforward to pass something that represents a BeamFnStateClient bound to a particular, instruction, transform, and window. I think this would make the code in FnApiDoFnRunner more clear as well.

Reply:
The issue is that because of continuation tokens we always will either be passing around a partial state request or partial state request builder because we need to fill in the continuation token which is what the `StateFetchingIterators.usingPartialRequestWithStateKey` is for.

Reply:
Now that I think about it more, we want both a supplier and a static context. The issue is that we want to use a supplier when being supplied to the DoFnContext for usage inside the DoFn, but when it is output outside of the DoFn, we need to have a snapshot of all the key components so that we can replay the requests. I would like to make this change to swap away from suppliers in a follow up change though.

Reply:
WDYT

Reply:
I added a test to cover that the side input is accessible after it has been output from the function.